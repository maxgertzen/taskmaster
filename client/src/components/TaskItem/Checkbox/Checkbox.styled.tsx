import styled from '@emotion/styled';

export const CheckboxContainer = styled.label({
  display: 'flex',
  alignItems: 'center',
  cursor: 'pointer',
  position: 'relative',
});

export const StyledCheckbox = styled.input<{ checked: boolean }>(
  ({ theme, checked }) => ({
    width: theme.spacing(2),
    height: theme.spacing(2),
    appearance: 'none',
    border: theme.borders.main(theme.colors.text),
    borderRadius: theme.spacing(0.5),
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    transition: 'all 150ms',
    cursor: 'pointer',
    color: theme.colors.text,
    '&::checked': {
      backgroundColor: theme.colors.primary,
    },
    '&::after': {
      content: checked ? `"âœ“"` : '""',
      fontSize: theme.spacing(1.5),
      color: theme.colors.text,
    },
    ':focus': {
      boxShadow: `0 0 0 1px ${theme.colors.primary}66`,
    },
    marginRight: theme.spacing(1.5),
  })
);

export const Label = styled.span(({ theme }) => ({
  marginLeft: theme.spacing(1),
  fontSize: theme.typography.body.fontSize,
}));
