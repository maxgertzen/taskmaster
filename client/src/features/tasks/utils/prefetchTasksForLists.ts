import { QueryClient } from '@tanstack/react-query';

import { QUERY_KEYS } from '@/shared/api/query-keys';
import { STALE_TIME } from '@/shared/constants/staleTime';

import { fetchTasks } from '../api';

export const prefetchTasksForLists = (
  listIds: string[],
  queryClient: QueryClient,
  token: string | null
) => {
  if (!token || !listIds.length) return;

  listIds.forEach((listId) => {
    queryClient.prefetchQuery({
      queryKey: QUERY_KEYS.tasks({ listId }),
      queryFn: fetchTasks(token),
      staleTime: STALE_TIME,
    });
  });
};
