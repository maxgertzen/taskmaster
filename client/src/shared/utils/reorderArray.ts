export const reorderArray = <T>(
  array: T[],
  fromIndex: number,
  toIndex: number
): T[] => {
  const reordered = [...array];
  const [movedItem] = reordered.splice(fromIndex, 1);
  reordered.splice(toIndex, 0, movedItem);
  return reordered;
};
