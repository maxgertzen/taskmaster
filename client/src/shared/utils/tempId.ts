export const generateTempId = (prefix: string = 'temp') =>
  `${prefix}-${Date.now()}`;
