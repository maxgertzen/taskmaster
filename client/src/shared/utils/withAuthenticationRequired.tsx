import { withAuthenticationRequired as realWithAuthenticationRequired } from '@auth0/auth0-react';

import { IS_AUTH0_DISABLED } from '@/constants';
import { mockWithAuthenticationRequired } from '@/mocks';

const isMock = IS_AUTH0_DISABLED === 'true';

export const withAuthenticationRequired = isMock
  ? mockWithAuthenticationRequired
  : realWithAuthenticationRequired;
