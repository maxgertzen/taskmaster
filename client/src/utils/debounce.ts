export function debounce<T extends (...args: Parameters<T>) => ReturnType<T>>(
  func: T,
  delay: number
): ((...args: Parameters<T>) => void) & { cancel: () => void } {
  let timeoutId: ReturnType<typeof setTimeout> | null = null;

  const debouncedFunction = (...args: Parameters<T>) => {
    if (timeoutId) clearTimeout(timeoutId);
    timeoutId = setTimeout(() => func(...args), delay);
  };

  debouncedFunction.cancel = () => {
    if (timeoutId) clearTimeout(timeoutId);
  };

  return debouncedFunction;
}
