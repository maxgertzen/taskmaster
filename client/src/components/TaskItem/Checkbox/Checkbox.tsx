import React from 'react';

import { CheckboxContainer, Label, StyledCheckbox } from './Checkbox.styled';

interface CheckboxProps {
  checked: boolean;
  onChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
  label?: string;
}

export const Checkbox: React.FC<CheckboxProps> = ({
  checked,
  onChange,
  label = '',
}) => {
  return (
    <CheckboxContainer>
      <StyledCheckbox type='checkbox' checked={checked} onChange={onChange} />
      {label && <Label>{label}</Label>}
    </CheckboxContainer>
  );
};
