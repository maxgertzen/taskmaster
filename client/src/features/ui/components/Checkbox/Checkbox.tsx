import React from 'react';

import { CheckboxContainer, Label, StyledCheckbox } from './Checkbox.styled';

interface CheckboxProps {
  checked: boolean;
  onChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
  label?: string;
}

export const Checkbox: React.FC<CheckboxProps> = ({
  onChange,
  checked = false,
  label = '',
}) => {
  const handleOnChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    onChange?.(e);
  };

  return (
    <CheckboxContainer>
      <StyledCheckbox
        data-testid='checkbox'
        type='checkbox'
        checked={checked}
        onChange={handleOnChange}
      />
      {label && <Label>{label}</Label>}
    </CheckboxContainer>
  );
};
