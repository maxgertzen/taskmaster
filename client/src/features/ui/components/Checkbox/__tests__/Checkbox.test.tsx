import { describe, it, expect, vi } from 'vitest';

import { render, screen } from '@/tests/utils';

import { Checkbox } from '../Checkbox';

describe('Checkbox', () => {
  it('renders correctly in light theme', () => {
    const { container } = render(
      <Checkbox checked={true} onChange={vi.fn()} label='Label' />,
      { theme: 'light' }
    );
    expect(container).toMatchSnapshot();
  });

  it('renders correctly in dark theme', () => {
    const { container } = render(
      <Checkbox checked={true} onChange={vi.fn()} label='Label' />,
      { theme: 'dark' }
    );
    expect(container).toMatchSnapshot();
  });

  it('renders the label', () => {
    const { container } = render(
      <Checkbox checked={true} onChange={vi.fn()} label='Label' />
    );
    expect(container).toMatchSnapshot();
  });

  it('calls the onChange function when clicked', async () => {
    const handleChange = vi.fn();
    const { user } = render(
      <Checkbox checked={true} onChange={handleChange} label='Label' />
    );

    const checkbox = await screen.findByTestId('checkbox');
    await user.click(checkbox);

    expect(handleChange).toHaveBeenCalledTimes(1);
  });

  it('is checked when checked prop is true', () => {
    render(<Checkbox checked={true} onChange={vi.fn()} label='Label' />);

    const checkbox = screen.getByTestId('checkbox');
    expect(checkbox).toBeChecked();
  });

  it('is not checked when checked prop is false', () => {
    render(<Checkbox checked={false} onChange={vi.fn()} label='Label' />);

    const checkbox = screen.getByTestId('checkbox');
    expect(checkbox).not.toBeChecked();
  });
});
