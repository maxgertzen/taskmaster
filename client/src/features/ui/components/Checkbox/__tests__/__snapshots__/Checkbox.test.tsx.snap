// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Checkbox > renders correctly in dark theme 1`] = `
<div>
  <label
    class="css-1kptg10"
  >
    <input
      checked=""
      class="css-1f6w2vg"
      data-testid="checkbox"
      type="checkbox"
    />
    <span
      class="css-9nuk9i"
    >
      Label
    </span>
  </label>
</div>
`;

exports[`Checkbox > renders correctly in light theme 1`] = `
<div>
  <label
    class="css-1kptg10"
  >
    <input
      checked=""
      class="css-1x1rvkf"
      data-testid="checkbox"
      type="checkbox"
    />
    <span
      class="css-9nuk9i"
    >
      Label
    </span>
  </label>
</div>
`;

exports[`Checkbox > renders the label 1`] = `
<div>
  <label
    class="css-1kptg10"
  >
    <input
      checked=""
      class="css-1x1rvkf"
      data-testid="checkbox"
      type="checkbox"
    />
    <span
      class="css-9nuk9i"
    >
      Label
    </span>
  </label>
</div>
`;
